src/prefect/events/worker.py

```diff
@@ -0,0 +1,138 @@
+import asyncio
+import atexit
+import concurrent.futures
+import threading
+from contextlib import asynccontextmanager, contextmanager
+from typing import Any, Dict, Optional, Type
+
+from prefect._internal.compatibility.experimental import experiment_enabled
+from prefect._internal.concurrency.event_loop import call_soon_in_loop
+from prefect.settings import PREFECT_API_KEY, PREFECT_API_URL, PREFECT_CLOUD_API_URL
+
+from .clients import EventsClient, NullEventsClient, PrefectCloudEventsClient
+from .schemas import Event
+
+
+class EventsWorker:
+    def __init__(
+        self,
+        client_type: Type[EventsClient],
+        client_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        self._client_type = client_type
+        self._client_kwargs = client_kwargs or {}
+        self._client: EventsClient
+
+        self._thread = threading.Thread(
+            target=self._start_loop, name="events-worker", daemon=True
+        )
+        self._ready_future = concurrent.futures.Future()
+        self._lock = threading.Lock()
+
+        self._loop: asyncio.AbstractEventLoop
+        self._queue: asyncio.Queue[Optional[Event]]
+
+        atexit.register(self.stop)
+
+    def _start_loop(self):
+        """Entrypoint for the thread."""
+        asyncio.run(self._main_loop())
+
+    async def _main_loop(self):
+        """Orchestrate the events client and emit any queued events."""
+        self._loop: asyncio.AbstractEventLoop = asyncio.get_running_loop()
+        self._queue: asyncio.Queue[Optional[Event]] = asyncio.Queue()
+
+        try:
+            self._client = self._client_type(**self._client_kwargs)
+            async with self._client:
+                self._ready_future.set_result(True)
+                while True:
+                    event = await self._queue.get()
+
+                    if event is None:
+                        break
+
+                    await self._client.emit(event)
+        except Exception as exc:
+            self._ready_future.set_exception(exc)
+
+    def _enqueue(self, value: Optional[Event]):
+        call_soon_in_loop(self._loop, self._queue.put_nowait, value)
+
+    def start(self):
+        """Start worker thread if not already started."""
+        if not self._thread.is_alive():
+            with self._lock:
+                self._thread.start()
```

Comment:
I think threads can't be started multiple times, is the lock doing anything here?

Reply:
https://docs.python.org/3/library/threading.html#threading.Thread.start

> It must be called at most once per thread object. It arranges for the object’s [run()](https://docs.python.org/3/library/threading.html#threading.Thread.run) method to be invoked in a separate thread of control.
>
> This method will raise a [RuntimeError](https://docs.python.org/3/library/exceptions.html#RuntimeError) if called more than once on the same thread object.

Reply:
Good question and probably not.